Add support for darwin builds

james-orcales opened the pull request:
Tested both dynamic and statically linked executables with `./test_darwin.sh`. All examples are working too. Makefiles have warnings but I don't know how you'd like to refactor it. Other than that, all makefile targets work.

```
$ make
LD build/umka
ld: warning: -s is obsolete
CC obj/umka_api_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_common_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_compiler_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_const_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_decl_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_expr_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_gen_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_ident_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_lexer_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_runtime_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_stmt_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_types_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
CC obj/umka_vm_dynamic.o
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-dynamiclib' [-Wunused-command-line-argument]
LD build/libumka.dylib
ld: warning: -s is obsolete
```

james-orcales commented on examples/3dcam/build_3dcam_darwin.sh:
```diff
@@ -0,0 +1,25 @@
+    -L/opt/homebrew/lib \
+    -L$PWD \
+    -lraylib -lumka \
+    -framework Cocoa -framework IOKit -framework CoreVideo -framework OpenGL \
```
My bad, I didn't try it without these flags. I removed `-framework Cocoa -framework IOKit -framework CoreVideo -framework OpenGL \` and it works.

vtereshkov commented on build_darwin.sh:
```diff
@@ -13,11 +13,11 @@ rm -f *.o
 
 clang $clangflags -c umka.c
-clang umka.o -o umka -L$PWD -lm -lumka -Wl,-rpath,'$ORIGIN'
+clang umka.o -o umka -L$PWD -lm -lumka -Wl,-rpath,'@loader_path'
```
@james-orcales I expect from a contributor an absolutely clear understanding of what is going on in their pull requests, the more so that I don't have macOS at hand and so cannot check any of your claims myself.

You're explaining `@loader_path` by "consistency". For me, consistency means that:
* There is a stable Linux build script that uses `$ORIGIN`. I wrote and checked it myself
* Other contributors wrote a macOS script that was almost identical to that for Linux. This was consistent 
* Now you're changing it to some `@loader_path` and cannot even explain the meaning of this change. This is hugely inconsistent

I expect the _minimal_ possible changes to the existing codebase that solve the problem. Not a single character without explanation. That will be consistent.

james-orcales commented:
I've reverted all `$ORIGIN` replacements and removed the homebrew prompt.

Regarding @loader_path:

This [blog post](https://www.mikeash.com/pyblog/friday-qa-2009-11-06-linking-and-install-names.html) is a great summary of how dynamic linking works on Darwin. `@loader_path` is the direct equivalent of `$ORIGIN`. Despite `$ORIGIN` working, it is an undocumented translation of the Darwin's dynamic linker. For explicitness, `@loader_path` is better as its officially documented under the `dyld` man pages. But again, if you prioritize the most minimal changes then that's that.

```
@loader_path/
      This variable is replaced with the path to the directory containing the mach-o binary which contains the load command using @loader_path. Thus, in every
      binary, @loader_path resolves to a different path, whereas @executable_path always resolves to the same path. @loader_path is useful as the load path for a
      framework/dylib embedded in a plug-in, if the final file system location of the plugin-in unknown (so absolute paths cannot be used) or if the plug-in is used
      by multiple applications (so @executable_path cannot be used). If the plug-in mach-o file is at /some/path/Myfilter.plugin/Contents/MacOS/Myfilter and a
      framework dylib file is at /some/path/Myfilter.plugin/Contents/Frameworks/Foo.framework/Versions/A/Foo, then the framework load path could be encoded as
      @loader_path/../Frameworks/Foo.framework/Versions/A/Foo and the Myfilter.plugin directory could be moved around in the file system and dyld will still be able
      to load the embedded framework.
```


Regarding homebrew:

I don't consider this optional convenience prompt for Darwin users as violating the zero-dependency principle. In the project README.md itself, umka installation via homebrew is mentioned already. But you are correct that for mirroring the other bulid scripts, this was uncalled for.

vtereshkov commented on build_darwin.sh:
```diff
@@ -13,7 +13,7 @@ rm -f *.o
 
 clang $clangflags -c $sourcefiles
-clang -shared -fPIC *.o -o libumka.dylib -lm -ldl
+clang -dynamiclib -install_name @loader_path/libumka.dylib -fPIC *.o -o libumka.dylib -lm -ldl
```
@james-orcales What does this change mean? Did this script work without it?

james-orcales replied:
> Also, if this -dynamiclib option is necessary, why am I seeing a lot of warnings in your log?

@vtereshkov  That is a separate issue with the makefiles. `build_darwin.sh` executes with zero warnings. 

```
~/code/umka-lang git:(darwin-build)
$ rm -rf ./umka_darwin && ./build_darwin.sh && ./test_darwin.sh
Build finished
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (25, 1): Identifier a is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (23, 1): Identifier d is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (63, 1): Shadowed identifier a
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (65, 1): Shadowed identifier b
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (65, 1): Identifier b is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (72, 1): Shadowed identifier r
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (72, 1): Identifier r is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (71, 1): Identifier p is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (81, 1): Identifier g2res is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (57, 1): Identifier x is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (52, 1): Identifier c is not used
Warning /Users/james-orcales/code/umka-lang/tests/gc.um (48, 1): Identifier r is not used
Warning /Users/james-orcales/code/umka-lang/tests/multret.um (29, 1): Identifier z is not used
Warning /Users/james-orcales/code/umka-lang/tests/multret.um (29, 1): Identifier y is not used
Warning /Users/james-orcales/code/umka-lang/tests/multret.um (29, 1): Identifier x is not used
Warning /Users/james-orcales/code/umka-lang/tests/tour.um (73, 1): Identifier y is not used
Warning /Users/james-orcales/code/umka-lang/tests/tour.um (23, 1): Identifier ErrFn is not used
Warning /Users/james-orcales/code/umka-lang/tests/tour.um (12, 1): Identifier IntPtr is not used
Warning /Users/james-orcales/code/umka-lang/tests/typeeq.um (42, 1): Identifier p4 is not used
Warning /Users/james-orcales/code/umka-lang/tests/typeeq.um (40, 1): Identifier p2 is not used
Warning /Users/james-orcales/code/umka-lang/tests/weakptr3.um (1, 1): Module std is not used
Warning /Users/james-orcales/code/umka-lang/tests/fwdtypes.um (1, 1): Module MapNode is not used
23378 / 24727 bytes
OK
Time elapsed: 0.201 s
Time elapsed: 0.423 s
Time elapsed: 2.520 s
Time elapsed: 0.382 s
Time elapsed: 0.839 s
Time elapsed: 1.728 s
462 / 462 bytes
OK
```

The warnings are specific to the Makefiles only. With the way Makefiles are set up, -dynamiclib flag is also passed to the "compile-only" steps here through DYNAMIC_CFLAGS. `clang` warns that `-dynamiclib` is unused in compilation as its meant for the linker only.

```
$(OBJ_PATH)/%_static.o: src/%.c
	@echo CC $@
	@mkdir -p -- $(dir $@)
	@$(CC) $(STATIC_CFLAGS) -o $@ -c $^

$(OBJ_PATH)/%_dynamic.o: src/%.c
	@echo CC $@
	@mkdir -p -- $(dir $@)
	@$(CC) $(DYNAMIC_CFLAGS) -o $@ -c $^
```

For example, here it wouldn't warn because linking also takes place.

```
$(UMKA_LIB_DYNAMIC): $(OBJS_DYNAMIC)
	@echo LD $@
	@mkdir -p -- $(BUILD_PATH)/include/
	@$(CC) $(DYNAMIC_CFLAGS) -o $(UMKA_LIB_DYNAMIC) $^ $(LDFLAGS)
	@cp $(APIS) $(BUILD_PATH)/include/
```

I did say these warnings are ignorable and that they can go away with a little refactoring but I leave the refactoring up to you. All make targets are working.

> Makefiles have warnings but I don't know how you'd like to refactor it. Other than that, all makefile targets work.

vtereshkov commented on examples/3dcam/build_3dcam_darwin.sh:
```diff
@@ -0,0 +1,12 @@
+
+clang 3dcam.c -o 3dcam \
+    -I/opt/homebrew/include \
+    -L/opt/homebrew/lib \
```
@james-orcales You removed the Homebrew installation. Then why are you still referring to the Homebrew folders? I think you should have copied the Raylib binaries to the current folder.

james-orcales commented on tests/lib/build_lib_darwin.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+clang -fPIC -O3 -c lib.c
+clang -dynamiclib -fPIC *.o -o lib.umi
```
They’re conceptually similar, but -dynamiclib produces a true .dylib with metadata like an install_name, while -shared creates a Mach-O bundle instead. Since -install_name requires -dynamiclib, it makes sense to use it consistently on macOS to guide Darwin users.

james-orcales commented on umka_darwin/examples/3dcam/build_3dcam_darwin.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
```
Oh I'm sorry I didn't know this got commited.
